compyle/array.py

```diff
@@ -624,6 +741,22 @@ def cumsum(ary, backend=None, out=None):
         return wrap_array(output, backend)
 
 
+def take_bool_elwise(i, condition, ary, cum_sum_ar, out_ar):
+    if condition[i]:
+        out_ar[cum_sum_ar[i]-1] = ary[i]
+
+
+def take_bool(ary, condition, backend=None):
+    if backend is None:
+        backend = ary.backend
+    cumsum_ar = cumsum(condition, backend=backend)
+    out_ar = ones(cumsum_ar[-1], ary.dtype, backend=backend)
+    take_bool_elwise_an = annotate(take_bool_elwise)
+    e = Elementwise(take_bool_elwise_an, backend=backend)
```

Comment:
This should be cached right?  Each time `take_bool` is called it is going to create a fresh elementwise, you should use some form of memoization as done in many places in compyle.